### PR TITLE
refactor sdk

### PR DESCRIPTION
**Pull request by ali-718**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -47,7 +47,7 @@ const testUrl =
 function App() {
   return (
     <div style={{ width: 50, height: 50 }}>
-      <ImageCrush url={testUrl} width={50} height={50} />
+      <ImageCrush onLoad={onLoad} onError={onError} url={testUrl} width={50} height={50} />
```
Comment (fawaz-ahmed) [inline README.md]: Add on load and on error only in props table

File: README.md [hunk 2]
```diff
@@ -73,7 +73,7 @@ const testUrl =
 function App() {
   return (
     <div style={{ width: 50, height: 50 }}>
-      <ImageCrushAdoptive url={testUrl} />
+      <ImageCrushAdoptive onLoad={onLoad} onError={onError} url={testUrl} />
```
Comment (fawaz-ahmed) [inline README.md]: Same here

File: README.md [hunk 3]
```diff
@@ -120,6 +120,8 @@ function App() {
   return (
     <div ref={useRef()} style={{ width: 50, height: 50 }}>
       <ImageCrush
+        onLoad={onLoad}
```
Comment (fawaz-ahmed) [inline README.md]: Same here

File: example/src/modules/ImageCrush.js [hunk 4]
```diff
@@ -1,13 +1,27 @@
...
 const baseServiceUrl = "https://service.assetcrush.com";
 
-const ImageCrush = ({ url, width, height, ...props }) => {
+const ImageCrush = ({ onError = () => null, onLoad = () => null, url, width, height, acEnv = 'production', ...props }) => {
   const [image, setImage] = useState("");
+  const [isError, setIsError] = useState(false)
 
-  useEffect(() => {
-    if (!url) return;
+  const handleError = (e) => {
```
Comment (fawaz-ahmed) [inline example/src/modules/ImageCrush.js]: Use callback

File: example/src/modules/ImageCrush.js [hunk 5]
```diff
@@ -1,13 +1,27 @@
...
+    onError(e);
+    setIsError(true)
+  }
+
+  const handleRetry = () => {
```
Comment (fawaz-ahmed) [inline example/src/modules/ImageCrush.js]: Use callback

File: example/src/modules/ImageCrush.js [hunk 6]
```diff
@@ -1,13 +1,27 @@
...
+    fetchImage()
+  }
+
+  const fetchImage = () => {
+
```
Comment (fawaz-ahmed) [inline example/src/modules/ImageCrush.js]: Use calback

File: example/src/modules/ImageCrush.js [hunk 7]
```diff
@@ -16,17 +30,30 @@ const ImageCrush = ({ url, width, height, ...props }) => {
...
-    fetch(imageUrl, { headers: { "assetcrush-key": key } })
-      .then((r) => r.blob())
+    fetch(imageUrl, { headers: { "assetcrush-key": key, 'ac-env': acEnv } })
+      .then((r) => {
+        onLoad(r.headers)
+        return r.blob()
+      })
       .then((d) => {
         if (typeof window === undefined) return;
+        if (!d.type.includes('image')) {
```
Comment (fawaz-ahmed) [inline example/src/modules/ImageCrush.js]: Why have we added this block ?

File: example/src/modules/ImageCrush.js [hunk 8]
```diff
@@ -16,17 +30,30 @@ const ImageCrush = ({ url, width, height, ...props }) => {
...
+    if (!url) return;
+    fetchImage()
   }, [url, width, height]);
 
-  if (!image) return null;
+  if (isError) return <ErrorRender handleRetry={handleRetry} />
+  if (!image) return <Spinner />
 
-  return <img src={image} {...props} />;
+  return <img className={'fade-in-Image'} src={image} {...props} />;
```
Comment (fawaz-ahmed) [inline example/src/modules/ImageCrush.js]: class name should come as prop and be able to override like other props

File: example/src/modules/ErrorRender/ErrorRender.js [hunk 9]
```diff
@@ -0,0 +1,13 @@
+import React from 'react'
+import './styles.css'
+
+export const ErrorRender = ({ handleRetry }) => {
+    return (
+        <button onClick={handleRetry}>
+            <span className="shadow"></span>
+            <span className="edge"></span>
+            <span className="front text"> retry
```
Comment (fawaz-ahmed) [inline example/src/modules/ErrorRender/ErrorRender.js]: Remove text, since text has to be localized

**Review by fawaz-ahmed — CHANGES_REQUESTED** [review]
See comments. Fade in still to decide

File: README.md (hunk 1 above)
Comment (assetcrush) [inline README.md]: Remove `onLoad={onLoad} onError={onError}` from Usage, only add in props table

File: README.md (hunk 2 above)
Comment (assetcrush) [inline README.md]: same here

File: example/src/modules/ErrorRender/ErrorRender.js [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+    const _handleRetry = useCallback(() => {
+        clearTimeout(handler.current)
+
+        handler.current = setTimeout(handleRetry, 1000);
+    }, [handleRetry])
+
+
+    if (Icon) return <div onClick={_handleRetry}>{Icon}</div>
+    return (
+        <span style={{color: reloadIconColor,fontSize: height / 2}} onClick={_handleRetry} className="reload">
```
Comment (assetcrush) [inline example/src/modules/ErrorRender/ErrorRender.js]: Indent please

File: README.md [hunk 11]
```diff
@@ -47,7 +56,7 @@ const testUrl =
 function App() {
   return (
     <div style={{ width: 50, height: 50 }}>
-      <ImageCrush url={testUrl} width={50} height={50} />
+      <ImageCrush  url={testUrl} width={50} height={50} />
```
Comment (assetcrush) [inline README.md]: Extra space befor `url`

File: README.md [hunk 12]
```diff
@@ -73,7 +82,7 @@ const testUrl =
 function App() {
   return (
     <div style={{ width: 50, height: 50 }}>
-      <ImageCrushAdoptive url={testUrl} />
+      <ImageCrushAdoptive  url={testUrl} />
```
Comment (assetcrush) [inline README.md]: Remove extra space before url

File: README.md [hunk 13]
```diff
@@ -90,6 +99,15 @@ export default App;
 | `url`    | `string` | yes      | image url       |
 | `width`  | `string` | no       | width of image  |
 | `height` | `string` | no       | height of image |
+| `onLoad` | `func` | no       | when image successfully loads |
+| `onError` | `func` | no       | when image fails to load |
+| `animated` | `bool` | no       | image fade in effect |
+| `reloadIconColor` | `string` | no       | color of reload icon if image fails |
+| `spinnerIcon` | `string` | no       | spinner component icon while image loads |
+| `spinnerColor` | `string` | no       | color of spinner |
+| `isSpinner` | `bool` | no       | render spinner while image loads or not |
```
Comment (assetcrush) [inline README.md]: isSpinner is not accurate, change it to `hideSpinner`

File: README.md [hunk 14]
```diff
@@ -90,6 +99,15 @@ export default App;
...
+| `onError` | `func` | no       | if image fails to load (returns error details) |
```
Comment (assetcrush) [inline README.md:129]: onError and onLoad are repeating

File: example/src/App.js [hunk 15]
```diff
@@ -8,7 +8,7 @@ const testUrl =
 
 function App() {
   return (
-    <div style={{ width: 250, height: 250 }}>
+    <div style={{ width: 50, height: 50 }}>
```
Comment (assetcrush) [inline example/src/App.js]: Better not change this width and height

File: example/src/modules/ImageCrush.js [hunk 16]
```diff
@@ -1,32 +1,90 @@
...
+    return (
+      <ErrorRender
+        width={width}
+        height={height}
+        reloadIconColor={reloadIconColor}
+        icon={reloadIcon}
+        handleRetry={handleRetry}
+      />
+    );
+  if (image)
```
Comment (assetcrush) [inline example/src/modules/ImageCrush.js]: is this condition correct ? maybe there should be a `!`

File: example/src/modules/ImageCrush.js [hunk 17]
```diff
@@ -1,32 +1,90 @@
...
+    return (
+      <>{isSpinner && <Spinner icon={spinnerIcon} color={spinnerColor} />}</>
```
Comment (assetcrush) [inline example/src/modules/ImageCrush.js]: Do we need this fragment, if not remove it

File: example/src/modules/Spinner/styles.css [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+  display: inline-block;
+  position: relative;
+  width: 80px;
+  height: 80px;
+}
+.lds-roller div {
+  animation: lds-roller 1.2s cubic-bezier(0.5, 0, 0.5, 1) infinite;
+  transform-origin: 40px 40px;
+}
+/* .lds-roller div:after {
```
Comment (assetcrush) [inline example/src/modules/Spinner/styles.css]: remove commented css

**Review by assetcrush — CHANGES_REQUESTED** [review]
See comments

File: README.md [hunk 19]
```diff
@@ -87,55 +117,27 @@ export default App;
...
-| `url`    | `string` | yes      | image url       |
-| `width`  | `string` | no       | width of image  |
-| `height` | `string` | no       | height of image |
+| `backgroundColor`    | `string` | no      | color of background layer of image       |
+| `url`  | `string` | yes       | image url   |
+| `debounce` | `number` | no       | image layout handler debounce |
+| `wrapperStyle` | `object` | no       | style object for outermost div |
 
-Any additional props are passed down to underlying `<img />` element.
+Any additional props are passed down to underlying `<ImageCrush />` component.
```
Comment (assetcrush) [inline README.md]: underlying is img component

File: README.md [hunk 20]
```diff
@@ -57,6 +66,27 @@ export default App;
...
+| `url`    | `string` | yes      | image url       |
+| `width`  | `string` | no       | width of image  |
+| `height` | `string` | no       | height of image |
+| `onLoad` | `func` | no       | when image successfully loads |
+| `onError` | `func` | no       | when image fails to load |
+| `animated` | `bool` | no       | image fade in effect |
+| `reloadIconColor` | `string` | no       | color of reload icon if image fails |
+| `spinnerIcon` | `string` | no       | spinner component icon while image loads |
+| `spinnerColor` | `string` | no       | color of spinner |
+| `isSpinner` | `bool` | no       | render spinner while image loads or not |
```
Comment (assetcrush) [inline README.md]: Please rename this to `hideSpinner`

File: README.md [hunk 21]
```diff
@@ -87,55 +117,27 @@ export default App;
 
 | Prop     | Type     | Required | Note            |
 | -------- | -------- | -------- | --------------- |
-| `url`    | `string` | yes      | image url       |
-| `width`  | `string` | no       | width of image  |
-| `height` | `string` | no       | height of image |
+| `backgroundColor`    | `string` | no      | color of background layer of image       |
+| `url`  | `string` | yes       | image url   |
+| `debounce` | `number` | no       | image layout handler debounce |
+| `wrapperStyle` | `object` | no       | style object for outermost div |
```
Comment (assetcrush) [inline README.md:120]: All the ImageCrush props should also be part of ImageCrushAdoptive component

File: example/src/modules/Spinner/styles.css [hunk 22]
```diff
@@ -0,0 +1,84 @@
...
+  display: inline-block;
+  position: relative;
+  width: 80px;
+  height: 80px;
+}
+.assetcrush-lds-roller div {
+  animation: assetcrush-lds-roller 1.2s cubic-bezier(0.5, 0, 0.5, 1) infinite;
+  transform-origin: 40px 40px;
+}
+/* .assetcrush-lds-roller div:after {
```
Comment (assetcrush) [inline example/src/modules/Spinner/styles.css]: remove commented css

**Review by assetcrush — CHANGES_REQUESTED** [review]
Still changes needed

File: example/src/modules/ImageCrush.js [hunk 23]
```diff
@@ -1,32 +1,90 @@
...
+    (e) => {
+      onError(e);
+      setIsError(true);
+    },
+    [setIsError]
+  );
+
+  const handleRetry = useCallback(() => {
+    fetchImage();
+  }, [setIsError]);
```
Comment (assetcrush) [inline example/src/modules/ImageCrush.js]: Dependency will be `fetchImage` and move this function below fecthImage

File: example/src/modules/Spinner/Spinner.js [hunk 24]
```diff
@@ -0,0 +1,38 @@
+import React from "react";
+import "./styles.css";
+
+const styles = (backgroundColor) =>  ({
```
Comment (assetcrush) [inline example/src/modules/Spinner/Spinner.js]: remove this, since not used

**Review by assetcrush — CHANGES_REQUESTED** [review]
Overall PR looks good, just some minor issues

**Review by assetcrush — APPROVED** [review]

**Pull request merged**